packages/block-library/src/search/view.js

```diff
@@ -66,6 +81,41 @@ const { actions } = store(
 					actions.closeSearchInput();
 				}
 			},
+			*updateSearch() {
+				const { ref } = getElement();
+				const { value } = ref;
+
+				// Don't navigate if the search didn't really change.
+				if ( value === state.search ) {
+					return;
+				}
+
+				state.search = value;
+
+				// Debounce the search by 300ms to prevent multiple navigations.
+				// We can do this by yielding a promise that resolves after 300ms and
+				// then bailing out if the search has changed.
+				yield new Promise( ( resolve ) => setTimeout( resolve, 300 ) );
```

Comment:
Oh, I see what you mean. You're saying that it's possible that a new keystroke
could arrive AFTER we had already checked if `value !== state.search` but before
`updateSearch()` finishes execution. In that case, the "previous" search would NOT be cancelled even though it's stale.

For the first version, a simple debounce should be enough (I think). It's just a
UX optimization after all. But I will take a look at your example to see if we
can add it!

Reply:
@cbravobernal might wanna chime in because we looked at it in our pair programming 🙂

Reply:
Although seems more complicated, the 'include abort' method seems much better. That way we save a bunch of requests, right?

Reply:
The function will _virtually always_ bail / abort if there is a new keystroke. That's because **after** `yield new Promise( ( resolve ) => setTimeout( resolve, 300 ) )`  we check if the state had changed in the meantime:

https://github.com/WordPress/gutenberg/blob/818bc96a30ab3d944d0c6514032c6df9dbcaed6a/packages/block-library/src/search/view.js#L99-L101

Because state is updated **synchronously** in the Interactivity API, this should virtually always result in the correct debouncing.

---

Now, what I was referring to [above](https://github.com/WordPress/gutenberg/pull/63147/files#r1803359838) (and why I said virtually always_) is this: There is a tiny chance that the new keystroke will arrive within the few miliseconds that **this** part of the function is executing:

https://github.com/WordPress/gutenberg/blob/818bc96a30ab3d944d0c6514032c6df9dbcaed6a/packages/block-library/src/search/view.js#L102-L115

which would be **after** we check if the state has updated but before the function had finished. But I don't see that as a huge problem, and I like the simplicity of the current solution 🙂 .

Reply:
I disagree that the way this is implemented now is simpler. Concurrency is very complex and this is simplified by limiting concurrency, not by allowing it and attempting to control concurrent conditions.

Relying on pre- and post- conditions to manage concurrent operations is error prone. I'd much rather only allow 1 operation and worry about superseding and limiting concurrency than try to allow concurrent operations and reason about the way things can change concurrently. Reasoning about how things should happen in 300ms if certain things have changed or not is very difficult to do.

I'll share a suggestion for changes here shortly.

Reply:
This is the full implementation I'd recommend here for this method:

```diff
diff --git before/packages/block-library/src/search/view.js after/packages/block-library/src/search/view.js
index 9bdc827506..66e9fea674 100644
--- before/packages/block-library/src/search/view.js
+++ after/packages/block-library/src/search/view.js
@@ -3,9 +3,8 @@
  */
 import { store, getContext, getElement } from '@wordpress/interactivity';
 
-const isEmpty = ( obj ) =>
-	[ Object, Array ].includes( ( obj || {} ).constructor ) &&
-	! Object.entries( obj || {} ).length;
+/** @type {( () => void ) | null} */
+let supersedePreviousSearch = null;
 
 const { state, actions } = store(
 	'core/search',
@@ -81,9 +80,9 @@ const { state, actions } = store(
 					actions.closeSearchInput();
 				}
 			},
-			*updateSearch() {
-				const { ref } = getElement();
-				const { value } = ref;
+
+			*updateSearch( e ) {
+				const { value } = e.target;
 
 				// Don't navigate if the search didn't really change.
 				if ( value === state.search ) {
@@ -93,16 +92,22 @@ const { state, actions } = store(
 				state.search = value;
 
 				// Debounce the search by 300ms to prevent multiple navigations.
-				// We can do this by yielding a promise that resolves after 300ms and
-				// then bailing out if the search has changed.
-				yield new Promise( ( resolve ) => setTimeout( resolve, 300 ) );
-				if ( value !== state.search ) {
+				supersedePreviousSearch?.();
+				const { promise, resolve, reject } = Promise.withResolvers();
+				const timeout = setTimeout( resolve, 300 );
+				supersedePreviousSearch = () => {
+					clearTimeout( timeout );
+					reject();
+				};
+				try {
+					yield promise;
+				} catch {
 					return;
 				}
 
 				const url = new URL( window.location.href );
-				if ( ! isEmpty( value ) ) {
-					state.search = value;
+				url.searchParams.delete( 'paged' );
+				if ( value ) {
 					url.searchParams.set( 'instant-search', value );
 				} else {
 					url.searchParams.delete( 'instant-search' );
```

Reply:
There's a tradeoff between the simplicity of these lines and their correctness. I'd like like to take a slightly more comples and slightly more correct approach. It may be that we disagree on where it's best to go in this tradeoff. It's unlikely to have a big impact because the problems have potential to appear in edge cases and in the future.

My point is that the execution is not properly limited. It is not debounced. Execution still continues, possibly multiple times, instead of being cancelled when superseded.

Here's a quick demonstration.

<details><summary>Apply this patch</summary>

```diff
diff --git before/packages/interactivity-router/src/index.ts after/packages/interactivity-router/src/index.ts
index 8a46d4ce35..f62ffbdaec 100644
--- before/packages/interactivity-router/src/index.ts
+++ after/packages/interactivity-router/src/index.ts
@@ -261,6 +261,11 @@ export const { state, actions } = store< Store >( 'core/router', {
 		 * @return  Promise that resolves once the navigation is completed or aborted.
 		 */
 		*navigate( href: string, options: NavigateOptions = {} ) {
+			console.log(
+				'navigate called with %d times, now: %s',
+				++calls,
+				href
+			);
 			const { clientNavigationDisabled } = getConfig();
 			if ( clientNavigationDisabled ) {
 				yield forcePageReload( href );
@@ -455,3 +460,4 @@ if ( globalThis.IS_GUTENBERG_PLUGIN ) {
 		);
 	}
 }
+let calls = 0;
```

</details>

Then open the search and quckly type a letter and delete it: `a`<kbd>DEL</kbd>`a`<kbd>DEL</kbd>. If you continue doing that the issue becomes more apparent.

What is the desired result? Nothing should happen until 300ms after the final keypress.

What actually happens? Depending on how quickly the inputs happened, multiple promises will queue up, 300ms will pass, and function execution will continue for all of them. Some of them will hit the equality check and return, others will pass the equality check and continue. **At this point, we must consider the implications of multiple executions of the rest of the function in rapid succession or concurrently.**

This is where I would much rather limit concurrency where it appears and not worry about it further.

If you've heard me out and disagree, go ahead and make a different tradeoff. I won't block this PR about this detail.

Reply:
I gave it some further thought, and I concede that you're completely right. The implementation is simply incorrect when relying on the post-condition. Thank you for standing your ground and challenging my assumptions 🙂.

Another way to demonstrate why the behavior is buggy is to increase the timeout to something like 3000ms. Then, it's easier to observe that the function calls are not debounced properly. They 

I'll implement your recommendation!

Reply:
Please, even though WordPress' frontend support is unclear, always check the support of all the browser features we add. Support for `Promise.withResolvers` is still [a bit low (87%)](https://caniuse.com/mdn-javascript_builtins_promise_withresolvers) compared to the rest of the features we use, which have at least 93% support.

We can start a conversation here about what percentage of browsers WordPress should support; however, this API is not critical for the block's functionality, so I think it's simpler to look for an alternative implementation that is better supported.

Reply:
`Promise.withResolvers` is not essential and should be straightforward to replace 👍